sdks/python/apache_beam/io/gcp/gcsio.py

```diff
@@ -827,79 +554,22 @@ def _start_upload(self):
     # The uploader by default transfers data in chunks of 1024 * 1024 bytes at
     # a time, buffering writes until that size is reached.
     try:
-      self.client.objects.Insert(self.insert_request, upload=self.upload)
+      self._client.objects.Insert(self._insert_request, upload=self._upload)
     except Exception as e:  # pylint: disable=broad-except
       logging.error('Error in _start_upload while inserting file %s: %s',
-                    self.path, traceback.format_exc())
-      self.upload_thread.last_error = e
+                    self._path, traceback.format_exc())
+      self._upload_thread.last_error = e
     finally:
-      self.child_conn.close()
+      self._child_conn.close()
 
-  def write(self, data):
-    """Write data to a GCS file.
+  def put(self, data):
+    self._conn.send_bytes(data.tobytes())
 
-    Args:
-      data: data to write as str.
-
-    Raises:
-      IOError: When this buffer is closed.
-    """
-    self._check_open()
-    if not data:
-      return
-    self.write_buffer.extend(data)
-    if len(self.write_buffer) > self.write_buffer_size:
-      self._flush_write_buffer()
-    self.position += len(data)
-
-  def flush(self):
-    """Flushes any internal buffer to the underlying GCS file."""
-    self._check_open()
-    self._flush_write_buffer()
-
-  def tell(self):
-    """Return the total number of bytes passed to write() so far."""
-    return self.position
-
-  def close(self):
-    """Close the current GCS file."""
-    if self.closed:
-      logging.warn('Channel for %s is not open.', self.path)
-      return
-
-    self._flush_write_buffer()
-    self.closed = True
-    self.conn.close()
-    self.upload_thread.join()
+  def finish(self):
+    self._conn.close()
+    # TODO(udim): Add timeout=DEFAULT_HTTP_TIMEOUT_SECONDS * 2 and check
+    # isAlive.
```

Comment:
<!--new_thread; commit:79815e6df1d49c271a6c44e37fb2a3c46be0aed5; resolved:0-->
Is the TODO content just an optimization, or is this a correctness issue?

Reply:
Correctness. The way it's called, join() may block forever. It currently works but there might be flows in which the thread never exists. Adding a timeout is important, but it might introduce a bug so I added this TODO instead.

Reply:
Are you worried about the network hanging?  That may already be mitigated by using `timeout=DEFAULT_HTTP_TIMEOUT_SECONDS` in the `httplib2.Http` client.

Reply:
It's just incorrect to wait forever without a timeout or some mechanism to interrupt the wait.